File: jung-algorithms/src/main/java/edu/uci/ics/jung/layout/util/VisRunnable.java
package edu.uci.ics.jung.layout.util;

import edu.uci.ics.jung.algorithms.util.IterativeContext;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * a {@code Runnable} object to pass to the {@code Thread} that will perform the relax function on a
 * graph layout
 *
 * @author Tom Nelson
 */
public class VisRunnable implements Runnable {

  private static final Logger log = LoggerFactory.getLogger(VisRunnable.class);
  private final IterativeContext iterativeContext;
  private long sleepTime = 10;
  private boolean stop = false;

  public VisRunnable(IterativeContext iterativeContext) {
    log.trace("created a VisRunnable {} for {}", hashCode(), iterativeContext);
    this.iterativeContext = iterativeContext;
  }

  public void stop() {
    log.trace("told {} to stop", this);
    stop = true;
  }

  @Override
  public void run() {
    while (!iterativeContext.done() && !stop) {
      try {
        iterativeContext.step();
        try {
          Thread.sleep(sleepTime);
        } catch (InterruptedException ex) {
        }
      } catch (Exception ex) {
        ex.printStackTrace();
      }
    }
    if (iterativeContext.done()) {
      log.trace("done here because {} is done", hashCode());
    }
    if (stop) {
      log.trace("done here because {} stop = {}", hashCode(), stop);
    }
  }
}

File: jung-algorithms/src/main/kotlin/edu/uci/ics/jung/layout/util/VisRunnable.kt
package edu.uci.ics.jung.layout.util

import edu.uci.ics.jung.algorithms.util.IterativeContext
import org.slf4j.LoggerFactory

/**
 * a [Runnable] object to pass to the [Thread] that will perform the relax function on a
 * graph layout
 *
 * @author Tom Nelson
 */
class VisRunnable(
  private val iterativeContext: IterativeContext
) : Runnable {

  private val sleepTime: Long = 10
  @Volatile
  private var stop: Boolean = false

  init {
    log.trace("created a VisRunnable {} for {}", hashCode(), iterativeContext)
  }

  fun stop() {
    log.trace("told {} to stop", this)
    stop = true
  }

  override fun run() {
    while (!iterativeContext.done() && !stop) {
      try {
        iterativeContext.step()
        try {
          Thread.sleep(sleepTime)
        } catch (ex: InterruptedException) {
        }
      } catch (ex: Exception) {
        ex.printStackTrace()
      }
    }
    if (iterativeContext.done()) {
      log.trace("done here because {} is done", hashCode())
    }
    if (stop) {
      log.trace("done here because {} stop = {}", hashCode(), stop)
    }
  }

  companion object {
    private val log = LoggerFactory.getLogger(VisRunnable::class.java)
  }
}
